Performance test for login and home page

suchcodemuchwow opened the pull request:
- Remove `reporting-dashboard` test from performance tests and add login page and home page tests to get
performance metrics.
- Set network latency (**KBN_TEST_NETWORK_LATENCY**), download throughput (**KBN_TEST_DOWNLOAD_THROUGHPUT**) and upload throughput (**KBN_TEST_UPLOAD_THROUGHPUT**) through environment variables and fallback to default _100ms_ latency for **network latency**, _5MB_ for **download throughput** and _1MB_ for **upload throughput**.

Commit:
refactor(fix typo): fix typo in reporting dashboard test name

Commit:
feat(test/functional/services/remote/webdriver.ts): support env vars for throttling and throughput

Set network latency(KBN_TEST_NETWORK_LATENCY), download throughput(KBN_TEST_DOWNLOAD_THROUGHPUT) and
upload throughput(KBN_TEST_UPLOAD_THROUGHPUT) through environment variables and fallback to default
100ms latency for network latency, 5MB for download throughput and 1MB for upload throughput.

Commit:
test(x-pack/test/performance/tests/home.ts): replace perf reporting-dashboard test w/ home page test

Remove reporting-dashboard test from performance tests and add login page and home page tests to get
performance metrics.

kibanamachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/5010)
* Commit: d79354073a2e68226647f59d743ee2902be31a31
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-117861/d79354073a2e68226647f59d743ee2902be31a31)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/5010#a1f63a5a-62c8-47f5-8a53-2b5c49a361fc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/5010/jobs/a1f63a5a-62c8-47f5-8a53-2b5c49a361fc/artifacts/67a9fd7c-9fa7-41f1-9d0e-c4163bcd9d4d) Default CI Group #<span></span>4 / lens app  lens dashboard tests should be able to add filters/timerange by clicking in XYChart

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @suchcodemuchwow

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/5010","number":5010,"commit":"d79354073a2e68226647f59d743ee2902be31a31"}],"number":5010}
buildkite-pr-comment-->

Commit:
refactor(test/functional/services/remote/network_profiles.ts): mv network consts to network_profiles

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/5370)
* Commit: 9b5ef5a35f1249009fa71040c179f95550061248
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-117861/9b5ef5a35f1249009fa71040c179f95550061248)

### Failed CI Steps
* [Default CI Group #1](https://buildkite.com/elastic/kibana-pull-request/builds/5370#04c8e161-547d-4b80-8aee-2b990ed5ffeb)
* [Checks](https://buildkite.com/elastic/kibana-pull-request/builds/5370#e863c25f-1349-473b-88a2-8fb71afd99b3)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/5370#04c8e161-547d-4b80-8aee-2b990ed5ffeb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/5370/jobs/04c8e161-547d-4b80-8aee-2b990ed5ffeb/artifacts/a3a632c2-394e-4eab-8d09-718a22386cc4) Default CI Group #<span></span>1 / APM API tests trial 8.0.0 correlations latency with data and opbeans-node args runs queries and returns results

### History
* :yellow_heart: [Build #5010](https://buildkite.com/elastic/kibana-pull-request/builds/5010) was flaky d79354073a2e68226647f59d743ee2902be31a31

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @suchcodemuchwow

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/5370","number":5370,"commit":"9b5ef5a35f1249009fa71040c179f95550061248"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/5010","number":5010,"commit":"d79354073a2e68226647f59d743ee2902be31a31"}],"number":5370}
buildkite-pr-comment-->

Commit:
fix(test/functional/services/remote/network_profiles.ts): change download/upload speeds kb/s -> b/s

suchcodemuchwow commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/5964)
* Commit: f64cf2852b1e6d5a2d46e27e6656cf5d30bdd2c8
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-117861/f64cf2852b1e6d5a2d46e27e6656cf5d30bdd2c8)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #5370](https://buildkite.com/elastic/kibana-pull-request/builds/5370) failed 9b5ef5a35f1249009fa71040c179f95550061248
* :yellow_heart: [Build #5010](https://buildkite.com/elastic/kibana-pull-request/builds/5010) was flaky d79354073a2e68226647f59d743ee2902be31a31

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

cc @suchcodemuchwow

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/5964","number":5964,"commit":"f64cf2852b1e6d5a2d46e27e6656cf5d30bdd2c8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/5370","number":5370,"commit":"9b5ef5a35f1249009fa71040c179f95550061248"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/5010","number":5010,"commit":"d79354073a2e68226647f59d743ee2902be31a31"}],"number":5964}
buildkite-pr-comment-->

elasticmachine commented:
Pinging @elastic/kibana-operations (Team:Operations)

LeeDr reviewed (approved):
LTGM - code and results review only

tylersmalley commented on x-pack/test/performance/tests/home.ts:
```diff
@@ -0,0 +1,24 @@
+
+  describe('Login', () => {
+    it('login and navigate to homepage', async () => {
+      await PageObjects.common.navigateToApp('login');
```
For these tests, I think we might want to avoid these helpers or maybe improve them? There is some crazy retry and sleeps in [here](https://github.com/elastic/kibana/blob/main/test/functional/page_objects/common_page.ts#L221-L298). I

suchcodemuchwow replied:
I agree with you that there are quite extensive usage of retry logic one in another. However I also felt like since this `PageObjects.common.navigateToApp` is pretty generic if we remove/change the retries and waits in that function there can be some tests starts to fail because of timeouts. I spent some time on weekend to understand what is the best way of verifying (waiting explicitly, implicitly or retries) to land a page or check existence of an element, what I found out is there is no silver bullet for solving these issues with some magical logic each use case needs to be evaluated separately. In our case I believe navigateToApp function is trying to be as defensive as possible by adding multiple retries inside of each other. I think it would be nice if @spalger also adds some comments on this before I change something.

mshustov replied:
Can we log a test user in asynchronously to minimize the number of side effects affecting the result of performance testing?

suchcodemuchwow replied:
Yes for sure we can try it out 👍🏼. However I'm not familiar how to add test user 😅 and what would be the difference between this login flow and with the test user in terms of side effects

spalger replied:
My opinion is that we should consider alternate ways to navigate to the page only if we determine that it's a problem when we're trying to get more stable metrics and we've determined from the data that the way we're navigating is causing issues. Navigation is complex, making an implementation that works just for the performance stuff shouldn't be too bad, but "fixing" the common page object is a bit of a minefield if you ask me. There are a lot of considerations... Just my opinion.

tylersmalley replied:
I am fine moving forward with this while we generate data on the deviation. At this point, we're still not certain if we will be using the FTR, or will need bare metal.

tylersmalley commented on x-pack/test/performance/tests/home.ts:
```diff
@@ -0,0 +1,24 @@
+
+      await retry.waitFor('login page visible', () => testSubjects.exists('loginSubmit'));
+
+      await PageObjects.security.login();
```
We will probably want to validate we landed on the home page after login.

tylersmalley commented on x-pack/test/performance/tests/home.ts:
```diff
@@ -0,0 +1,24 @@
+    it('login and navigate to homepage', async () => {
+      await PageObjects.common.navigateToApp('login');
+
+      await retry.waitFor('login page visible', () => testSubjects.exists('loginSubmit'));
```
Do we need to use the `retry.waitFor` here? Can we instead supply `testSubjects.exists('loginSubmit')` with a timeout?

Commit:
refactor(x-pack/test/performance/tests/home.ts): check if we're at home after login, remove retry

- Remove retry logic while checking login page load
- Check if we are at home page after login to
Kibana

suchcodemuchwow commented:
@elasticmachine merge upstream

mshustov commented on test/functional/services/remote/network_profiles.ts:
```diff
@@ -0,0 +1,31 @@
+
+// Download (kb/s)	Upload (kb/s)	Latency (ms)
+// https://gist.github.com/theodorosploumis/fd4086ee58369b68aea6b0782dc96a2e
+export const NETWORK_PROFILES: { [key: string]: NetworkOptions } = {
```
optional nit: we can improve readability and autocomplete by:
```typescript
const sec = 1_000;
const kB = 1024;

export const NETWORK_PROFILES = {
  DEFAULT: { DOWNLOAD: 5 * kB * sec, UPLOAD: 1 * kB * sec, LATENCY: 0.1 * sec },
  ...
} as const;
```

mshustov commented on test/functional/services/remote/webdriver.ts:
```diff
@@ -288,14 +289,23 @@ async function attemptToCreateCommand(
   const { session, consoleLog$ } = await buildDriverInstance();
 
   if (throttleOption === '1' && browserType === 'chrome') {
+    const profile = process.env.KBN_NETWORK_TEST_PROFILE ?? 'DEFAULT';
```
nit: 
```js
if (!profile in NETWORK_PROFILES) {
  throw new Error(`Expected one of valid NETWORK_PROFILES, but given [${profile}]`);
```

mshustov commented on test/functional/services/remote/webdriver.ts:
```diff
@@ -288,14 +289,23 @@ async function attemptToCreateCommand(
+      `NETWORK THROTTLED with profile ${profile}: ${downloadThroughput}kbps down, ${uploadThroughput}kbps up, ${latency} ms latency.`
+    );
 
     (session as any).setNetworkConditions({
```
optional nit: it's not related to the current PR but let's use `@ts-expect-error` to detect a moment when selenium webdriver typings add `setNetworkConditions` declaration.

mshustov commented on x-pack/test/performance/tests/index.ts:
```diff
@@ -11,6 +11,6 @@ export default function ({ loadTestFile }: FtrProviderContext) {
     this.tags('ciGroup8');
 
-    loadTestFile(require.resolve('./reporting_dashboard'));
+    loadTestFile(require.resolve('./home'));
```
Does it mean that `x-pack/test/performance/tests/reporting_dashboard.ts` test suit can be removed?

suchcodemuchwow replied:
Yes

tylersmalley replied:
Should be able to delete that file as part of this PR

Commit:
refactor: refactor based on pr comments

- Make network profiles more readable
- Use @ts-expect-error to detect a moment when selenium webdriver typings add setNetworkConditions declaration
- Check environment variable if it exist in NETWORK_PROFILES

kibanamachine committed:
Merge branch 'main' into 2021-11-08-performance-test-login-and-home-page

kibanamachine commented:
The following labels were identified as gaps in your version labels and will be added automatically:
- v8.1.0

If any of these should not be on your pull request, please manually remove them.

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.0](https://github.com/elastic/kibana/pull/119088)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/119088">](https://github.com/elastic/kibana/pull/119088) |
| ✅ |  [7.16](https://github.com/elastic/kibana/pull/119089)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/119089">](https://github.com/elastic/kibana/pull/119089) |

The backport PRs will be merged automatically after passing CI.

lizozom commented:
@suchcodemuchwow does this resolve https://github.com/elastic/kibana/issues/103940 ?

If so, could you please close it?
